src/tensor_ops/convtrans2d/cuda_kernel.rs

```diff
@@ -0,0 +1,180 @@
+use cudarc::cublas::{CudaBlas, Gemm};
+use cudarc::driver::{DeviceRepr, DeviceSlice, LaunchAsync, LaunchConfig, ValidAsZeroBits};
+
+use crate::tensor_ops::matmul::cuda_kernel::sgemm_batch;
+use crate::{
+    shapes::*,
+    tensor::{Cuda, Tensor},
+};
+
+use std::sync::Arc;
+
+unsafe impl DeviceRepr for super::ConvTrans2DOp {}
+
+const PTX_SRC: &str = include_str!(concat!(env!("OUT_DIR"), "/convtrans2d.ptx"));
+
+trait HasCudaKernel<E> {
+    const MOD: &'static str;
+    const FNS: &'static [&'static str];
+}
+
+impl HasCudaKernel<f32> for Cuda {
+    const MOD: &'static str = "convtrans2d_f32";
+    const FNS: &'static [&'static str] = &[
+        "unfold_input_into_patches_f32",
+        "unfold_output_into_patches_f32",
+        "transpose_and_broadcast_filters_f32",
+        "sum_transposed_filters_f32",
+    ];
+}
+
+impl HasCudaKernel<f64> for Cuda {
+    const MOD: &'static str = "convtrans2d_f64";
+    const FNS: &'static [&'static str] = &[
+        "unfold_input_into_patches_f64",
+        "unfold_output_into_patches_f64",
+        "transpose_and_broadcast_filters_f64",
+        "sum_transposed_filters_f64",
+    ];
+}
+
+fn make_4d<S: Shape>(strides: S::Concrete) -> [usize; 4] {
+    match S::NUM_DIMS {
+        3 => [0, strides[0], strides[1], strides[2]],
+        4 => [strides[0], strides[1], strides[2], strides[3]],
+        _ => unreachable!("Only implemented for 3d & 4d arrays"),
+    }
+}
+
+impl<E: Dtype + ValidAsZeroBits> super::ConvTrans2DKernel<E> for Cuda
+where
+    Self: HasCudaKernel<E>,
+    CudaBlas: Gemm<E>,
+{
+    fn forward<L: Shape, R: Shape, O: Shape>(
+        &self,
+        op: super::ConvTrans2DOp,
+        lhs: &Tensor<L, E, Self>,
+        rhs: &Tensor<R, E, Self>,
+        out: &mut Tensor<O, E, Self>,
+    ) -> Result<(), Self::Err> {
+        if !self.dev.has_func(Self::MOD, Self::FNS[0]) {
+            self.dev.load_ptx(PTX_SRC.into(), Self::MOD, Self::FNS)?;
+        }
+
+        let patches_numel = op.batch * op.chan_in * op.kernel * op.kernel * op.h_out * op.w_out;
+        let mut patches = self.dev.alloc_zeros::<E>(patches_numel)?;
+        let img_strides = self.dev.htod_copy(make_4d::<L>(lhs.strides).into())?;
+        let unfold_fn = self.dev.get_func(Self::MOD, Self::FNS[0]).unwrap();
+        let cfg = LaunchConfig::for_num_elems(patches.len() as u32);
+        let params = (op, lhs.data.as_ref(), &img_strides, &mut patches);
+        unsafe { unfold_fn.launch(cfg, params) }?;
+
+        // (O, C * K * K) * (B, C * K * K, OH * OW) = (B, O, OH * OW)
+        let m = op.chan_out;
+        let k = op.chan_in * op.kernel * op.kernel;
+        let n = op.h_out * op.w_out;
+        unsafe {
+            sgemm_batch(
+                self.blas.as_ref(),
```

Comment:
This was recently changed (sorry about that!). Can you merge main into this branch?
```suggestion
            self.gemm_batch(
```